pkg/vsphere/vm/vm.go

```diff
@@ -731,3 +734,149 @@ func (vm *VirtualMachine) RemoveSnapshotByRef(ctx context.Context, snapshot *typ
 
 	return object.NewTask(vm.Vim25(), res.Returnval), nil
 }
+
+// CheckHostCompatibility checks that a host is compatible for VM placement.
+func (vm *VirtualMachine) CheckHostCompatibility(ctx context.Context, host *types.ManagedObjectReference) (*object.Task, error) {
+	req := types.CheckCompatibility_Task{
+		This: vm.ServiceContent.VmCompatibilityChecker.Reference(),
+		Vm:   vm.Reference(),
+		Host: host,
+	}
+
+	res, err := methods.CheckCompatibility_Task(ctx, vm.Vim25(), &req)
+	if err != nil {
+		return nil, err
+	}
+	return object.NewTask(vm.Vim25(), res.Returnval), nil
+}
+
+// PowerOn powers on a VM. If the environment is VC without DRS enabled, it will attempt to relocate  the VM
+// to the most suitable host in the cluster. If relocation or subsequent power-on fail, it will attempt the next
+// best host, and repeat this process until a successful power-on is achieved or there are no more hosts to try.
+func (vm *VirtualMachine) PowerOn(op trace.Operation) error {
+	drs := vm.Session.DRSEnabled != nil && *vm.Session.DRSEnabled
+	cls := vm.InCluster(op)
+
+	op.Debugf("DRS enabled: %t", drs)
+	op.Debugf("%s resides in a cluster: %t", vm.Reference.String(), cls)
+
+	// otherwise place the VM before powering it on
+	hmp := performance.NewHostMetricsProvider(vm.Session.Vim25())
+	rhp, err := placement.NewRankedHostPolicy(op, vm.Cluster, hmp)
+	if err != nil {
+		return err
+	}
+
+	// any of the following conditions are sufficient to power-on the VM:
+	// - environment is not VC
+	// - DRS is enabled
+	// - the VM does not belong to a cluster
+	// - the current host is adequate for power-on
+	if !vm.IsVC() || drs || !cls || rhp.CheckHost(op, vm.VirtualMachine) {
+		return vm.powerOn(op)
+	}
+
+	op.Warnf("Host is not adequate for power-on, getting placement recommendation")
+
+	var hosts, subset []*object.HostSystem
+
+	f := func() error {
+		hosts, err = rhp.RecommendHost(op, subset)
+		if err != nil {
+			return fmt.Errorf("error recommending host: %s", err.Error())
+		}
+		return nil
+	}
+
+	conf := retry.NewBackoffConfig()
+	conf.MaxInterval = 1 * time.Second
+	conf.MaxElapsedTime = 20 * time.Second
+
+	err = retry.DoWithConfig(f, retry.OnError, conf)
+	if err != nil {
+		return err
+	}
+
+	for len(hosts) > 0 {
+		op.Debugf("hosts: %s", hosts)
+		op.Infof("Placement recommended %s", hosts[0].Reference().String())
+
+		if vm.relocate(op, hosts[0]) != nil {
+			op.Warnf("VM relocation failed: %s", err.Error())
+		} else if vm.powerOn(op) != nil {
+			op.Warnf("VM power-on failed: %s", err.Error())
+		} else {
+			return nil
+		}
+
+		// if relocation or powerOn fails, something has changed and we need a new recommendation
+		subset = hosts[1:]
```

Comment:
So we essentially limit the number of retries by the number of hosts in the cluster?

Reply:
Should we subset This based on the number of hosts? if something changes, it is does not necessarily mean that the provided host won't be a valid host in a future recommendation. Why not limit this to a X number of tries for host placement? Or X seconds of attempted placement before failure? I notice that we are retrying in the case of any error, so avoiding things like host communication errors could be an issue. But I would imagine that the recommendation engine not being able to communicate with a host automatically invalidates it as an option.

Reply:
If we get here, it basically means that things were ok when we got a recommendation, but things are not ok by the time we attempt power on. At this point, we need to attempt a new recommendation. If we get to the end and there are no more hosts, we return error and could retry at the `vm.PowerOn` call (see my previous comment). After reading your comments, I'm starting to think that may be a better approach.

Reply:
Lord @hickeng, I summon thee! Come forth and weigh in on this matter.

Reply:
Yeah I mean worse case, the host is not recommended anymore. Best case we get a new recommendation(could be same host or different) and that host works. Or we try again, if the cluster is experiencing some extreme problems then the host list could return none(way over provisioned, or total Host blackout... #doomsday). Or you hit a timelimit or attempt limit.

Reply:
If we are talking a potential different retry medium. I would think going with an attempt limit would be best since it avoids extremely high latency networks(though could lead to long start times). :)

Reply:
@matthewavery by "something changing" I mean something like the host going into maintenance mode or getting disconnected, which (obviously) would certainly invalidate it as a choice. Suddenly becoming a less ideal host won't affect anything as currently implemented.

We have a few options here: we can retry at the very top level entrypoint (`vm.PowerOn()`), we can retry on attempts to relocate and/or actually power on the VM, we can retry on host metrics gathering, or we can retry on host recommendation (or some combination of some or all of the above). We are definitely not looking to over-engineer this, though, so I'm reluctant to even go down this road, at least for now.